interop-testing/src/test/java/io/grpc/testing/integration/MoreInProcessTest.java

```diff
@@ -0,0 +1,269 @@
+public class MoreInProcessTest {
+  private static final String UNIQUE_SERVER_NAME =
+      "in-process server for " + MoreInProcessTest.class;
+  @Rule
+  public final Timeout globalTimeout = new Timeout(1000);
+  // use a mutable service registry for later registering the service impl for each test case.
+  private final MutableHandlerRegistry serviceRegistry = new MutableHandlerRegistry();
+  private final Server inProcessServer = InProcessServerBuilder.forName(UNIQUE_SERVER_NAME)
+      .fallbackHandlerRegistry(serviceRegistry).directExecutor().build();
+  private final ManagedChannel inProcessChannel =
+      InProcessChannelBuilder.forName(UNIQUE_SERVER_NAME).directExecutor().build();
+
+  @Before
+  public void setUp() throws Exception {
+    inProcessServer.start();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    inProcessChannel.shutdownNow();
+    inProcessServer.shutdownNow();
+  }
+
+  @Test
+  public void asyncClientStreaming_serverResponsePriorToRequest() throws Exception {
+    // implement a service
+    final StreamingInputCallResponse fakeResponse =
+        StreamingInputCallResponse.newBuilder().setAggregatedPayloadSize(100).build();
+    TestServiceImplBase clientStreamingImpl = new TestServiceImplBase() {
+      @Override
+      public StreamObserver<StreamingInputCallRequest> streamingInputCall(
+          StreamObserver<StreamingInputCallResponse> responseObserver) {
+        // send response directly
+        responseObserver.onNext(fakeResponse);
+        responseObserver.onCompleted();
+        return new StreamObserver<StreamingInputCallRequest>() {
+          @Override
+          public void onNext(StreamingInputCallRequest value) {
+          }
+
+          @Override
+          public void onError(Throwable t) {
+          }
+
+          @Override
+          public void onCompleted() {
+          }
+        };
+      }
+    };
+    serviceRegistry.addService(clientStreamingImpl);
+    // implement a client
+    final CountDownLatch finishLatch = new CountDownLatch(1);
+    final AtomicReference<StreamingInputCallResponse> responseRef =
+        new AtomicReference<StreamingInputCallResponse>();
+    final AtomicReference<Throwable> throwableRef =
+        new AtomicReference<Throwable>();
+    StreamObserver<StreamingInputCallResponse> responseObserver =
+        new StreamObserver<StreamingInputCallResponse>() {
+          @Override
+          public void onNext(StreamingInputCallResponse response) {
+            responseRef.set(response);
+          }
+
+          @Override
+          public void onError(Throwable t) {
+            throwableRef.set(t);
+            finishLatch.countDown();
+          }
+
+          @Override
+          public void onCompleted() {
+            finishLatch.countDown();
+          }
+        };
+
+    // make a gRPC call
+    TestServiceGrpc.newStub(inProcessChannel).streamingInputCall(responseObserver);
+
+    assertTrue(finishLatch.await(1, TimeUnit.SECONDS));
```

Comment:
This is a strange timeout when there is a overall `Timeout` of one second. Either this timeout should be smaller or the overall timeout should be higher.